packages/eslint-plugin/src/rules/no-non-null-asserted-nullish-coalescing.ts

```diff
@@ -0,0 +1,94 @@
+import { TSESTree, TSESLint } from '@typescript-eslint/experimental-utils';
+import {
+  Definition,
+  DefinitionType,
+  Variable,
+} from '@typescript-eslint/scope-manager';
+import * as util from '../util';
+
+function hasAssignmentBeforeNode(
+  variable: Variable,
+  node: TSESTree.Node,
+): boolean {
+  return (
+    variable.references.some(
+      ref => ref.isWrite() && ref.identifier.range[1] < node.range[1],
+    ) ||
+    variable.defs.some(
+      def =>
+        isDefinitionWithAssignment(def) && def.node.range[1] < node.range[1],
+    )
+  );
+}
+
+function isDefinitionWithAssignment(definition: Definition): boolean {
+  if (definition.type !== DefinitionType.Variable) {
+    return false;
+  }
+
+  const variableDeclarator = definition.node;
+  return variableDeclarator.definite ?? variableDeclarator.init !== null;
```

Comment:
Whilst your check is technically correct as we never emit a `definite: false`, let's future-proof this as we eventually want to remove these "true or null" properties.

```suggestion
  return variableDeclarator.definite === true || variableDeclarator.init !== null;
```